substrate/frame/revive/src/evm/block_hash/block_builder.rs

```diff
@@ -0,0 +1,494 @@
+		}
+	}
+
+	/// Store the first transaction and receipt in pallet storage.
+	fn pallet_put_first_values(&mut self, values: (Vec<u8>, Vec<u8>)) {
+		crate::EthBlockBuilderFirstValues::<T>::put(Some(values));
+	}
+
+	/// Take the first transaction and receipt from pallet storage.
+	fn pallet_take_first_values(&mut self) -> Option<(Vec<u8>, Vec<u8>)> {
+		crate::EthBlockBuilderFirstValues::<T>::take()
+	}
+
+	/// Process a single transaction at a time.
+	pub fn process_transaction(
+		&mut self,
+		transaction_encoded: Vec<u8>,
+		success: bool,
+		gas_used: Weight,
+		encoded_logs: Vec<u8>,
+		receipt_bloom: LogsBloom,
+	) {
+		let tx_hash = H256(keccak_256(&transaction_encoded));
+		self.tx_hashes.push(tx_hash);
+
+		// Update the transaction trie.
+		let transaction_type = Self::extract_transaction_type(transaction_encoded.as_slice());
+
+		// Update gas and logs bloom.
+		self.gas_used = self.gas_used.saturating_add(gas_used.ref_time().into());
+		self.logs_bloom.accrue_bloom(&receipt_bloom);
+
+		// Update the receipt trie.
+		let encoded_receipt = AccumulateReceipt::encoded_receipt(
+			encoded_logs,
+			receipt_bloom,
+			success,
+			self.gas_used.as_u64(),
+			transaction_type,
+		);
+
+		self.gas_info.push(ReceiptGasInfo { gas_used: gas_used.ref_time().into() });
+
+		// The first transaction and receipt are returned to be stored in the pallet storage.
+		// The index of the incremental hash builders already expects the next items.
+		if self.tx_hashes.len() == 1 {
+			log::debug!(target: LOG_TARGET, "Storing first transaction and receipt in pallet storage");
+			self.pallet_put_first_values((transaction_encoded, encoded_receipt));
+			return;
+		}
+
+		if self.transaction_root_builder.needs_first_value(BuilderPhase::ProcessingValue) {
+			if let Some((first_tx, first_receipt)) = self.pallet_take_first_values() {
+				log::debug!(target: LOG_TARGET, "Loaded first transaction and receipt from pallet storage");
+				self.transaction_root_builder.set_first_value(first_tx);
+				self.receipts_root_builder.set_first_value(first_receipt);
+			} else {
+				log::error!(target: LOG_TARGET, "First transaction and receipt must be present at processing phase");
+			}
+		}
+
+		self.transaction_root_builder.add_value(transaction_encoded);
+		self.receipts_root_builder.add_value(encoded_receipt);
+	}
+
+	/// Build the ethereum block from provided data.
+	pub fn build(
+		&mut self,
+		block_number: U256,
+		parent_hash: H256,
+		timestamp: U256,
+		block_author: H160,
+		gas_limit: U256,
+	) -> (Block, Vec<ReceiptGasInfo>) {
+		if self.transaction_root_builder.needs_first_value(BuilderPhase::Build) {
+			if let Some((first_tx, first_receipt)) = self.pallet_take_first_values() {
+				self.transaction_root_builder.set_first_value(first_tx);
+				self.receipts_root_builder.set_first_value(first_receipt);
+			} else {
+				log::error!(target: LOG_TARGET, "First transaction and receipt must be present at build phase");
```

Comment:
In case of empty blocks this log will be printed.
I guess we'd prefer to avoid it.